### PR TITLE
Boxing conversion

### PR DESCRIPTION
**Pull request by mpalat**
## What it does
Boxing part of issue #2298 
## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 1]
```diff
@@ -242,6 +245,12 @@ private void generateTypeCheck(BlockScope scope, CodeStream codeStream) {
 	}
 }
 
+private void setPatternIsTotalType() {
+	if (this.pattern != null) {
+		this.pattern.isTotalTypeNode = true;
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:250]: does `isTotalTypeNode` implement what 14.11.1.1 calls "A set of case elements, PCE, covers a type T"?

If so, then I don't think setting true is correct for any paths that use narrowing, IOW, when a conversion is not "unconditionally exact".

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 2]
```diff
@@ -346,25 +362,29 @@ public TypeBinding resolveType(BlockScope scope) {
...
-	} else {
+
+		// Do Nothing - no additional steps required for conversion later.
+
+	} else if (route == PrimitiveConversionRoute.BOXING_CONVERSION
+			|| route == PrimitiveConversionRoute.BOXING_CONVERSION_AND_WIDENING_REFERENCE_CONVERSION) {
 		addSecretExpressionValue(scope, expressionType);
+	} else if (route == PrimitiveConversionRoute.NO_CONVERSION_ROUTE) {
+		if (flagError)
+			scope.problemReporter().notCompatibleTypesError(this, expressionType, checkedType);
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:387]: I find it difficult to see, why some invocations should check these conversions and not report an error. I'll propose a small change that avoids calling checkForPrimitives() if no primitives are in sight, which then allows removing `flagError`.

**Comment by stephan-herrmann** [comment]
@mpalat I had difficulties seeing which combinations of types are covered by the huge number of tests. In order to find out I started to fold many of the plain-numbered tests into few generic implementations. I'm aware that the generic tests are more complex, but OTOH those should help to see what is the focus of each test. I hope you don't mind.

As a result I could convince myself that identity and widening primitive conversions are fully covered.

With your consent I might start giving the same treatment to more groups of tests.

**Comment by stephan-herrmann** [comment]
> With your consent I might start giving the same treatment to more groups of tests.

I went ahead anyway, but put it into a fresh new PR: #2790

**Comment by mpalat** [comment]
> @mpalat I had difficulties seeing which combinations of types are covered by the huge number of tests. In order to find out I started to fold many of the plain-numbered tests into few generic implementations. I'm aware that the generic tests are more complex, but OTOH those should help to see what is the focus of each test. I hope you don't mind.
> 
> As a result I could convince myself that identity and widening primitive conversions are fully covered.
> 
> With your consent I might start giving the same treatment to more groups of tests.

@stephan-herrmann please feel free.. thanks

**Comment by stephan-herrmann** [comment]
@mpalat in the interest of time and while waiting for @srikanth-sankaran to weigh in in #2790, I suggest you might continue from https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2549/commits/432b4118f9dc81950c0876f52e582f506de095c0 from this PR, OK?

**Comment by stephan-herrmann** [comment]
Some code changes have been merged via other PRs, comments on the code are (a) moved to #2880, (b) no longer relevant.

Re-organized tests have been merged *additionally* in #2878

We're done.

**Pull request closed without merging**